File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2016/Y16D09.kt
package org.gristle.pdxpuzzles.advent.y2016

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.parsing.getIntList

class Y16D09(private val input: String) : Day {

    // regex to find markers in the string
    private val markerRx = Regex("""\(\d+x\d+\)""")

    // returns the decompressed length of a string. "recursive" boolean flag used for part 2.
    private fun String.decompressedLength(recursive: Boolean): Long {
        var decompressedLength = 0L

        // index advances to parse the string
        var index = 0

        // keep going until there are no more chars to parse
        while (index < length) {

            // get the next marker. If none found, return early, adding the length of the remaining unparsed string
            val marker = markerRx
                .find(substring(index))
                ?: return decompressedLength + substring(index).length // return early if no marker found

            // add the length of any characters preceding the marker
            decompressedLength += substring(index, index + marker.range.first).length

            // get the length and number of repeats in the marker
            val (sequenceLength, repeats) = marker.value.getIntList()

            // get the length of the sequence affected by the marker
            val sequence = substring(index + marker.range.last + 1, index + marker.range.last + 1 + sequenceLength)
                .let {

                    // If part 2, feed the sequence into the function recursively, otherwise just use the length of
                    // the string.
                    @Suppress("KotlinConstantConditions")
                    if (recursive) it.decompressedLength(recursive) else it.length.toLong()
                }.times(repeats)

            // add this length to the overall decompressed length
            decompressedLength += sequence

            // advance the index to the character after the sequence
            index += marker.range.last + sequenceLength + 1
        }
        return decompressedLength
    }

    override fun part1() = input.decompressedLength(false)
    override fun part2() = input.decompressedLength(true)
}

fun main() = Day.runDay(Y16D09::class)

//    Class creation: 13ms
//    Part 1: 110346 (2ms)
//    Part 2: 10774309173 (15ms)
//    Total time: 31ms

File: rust/advent/src/bin/2016/y16d09.rs
use advent::utilities::get_input::get_input;
use itertools::Itertools;
use lazy_regex::regex;
use utilities::{parsing::get_numbers::ContainsNumbers, structs::stopwatch::{ReportDuration, Stopwatch}};

type Input<'a> = &'a str;
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(16, 9).unwrap();
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn solve(data: Input, recursive: bool) -> usize {
    let marker_rx = regex!(r"\(\d+x\d+\)");
    let mut length = 0;
    let mut index = 0;
    while index < data.len() {
        // get the next marker. If none found, return early, adding the length of the remaining unparsed string
        let Some(marker) = marker_rx.find(&data[index..]) else {
            return length + data[index..].len();
        };

        // add the length of any characters preceding the marker
        length += data[index..index + marker.start()].len();

        // get the length and number of repeats in the marker
        let (sequence_length, repeats): (usize, usize) = marker.as_str().get_numbers().collect_tuple().unwrap();

        // get the length of the sequence affected by the marker
        let post_marker = index + marker.end();
        let sequence = &data[post_marker..post_marker + sequence_length];
        let sequence = if recursive {
            solve(&sequence, recursive)
        } else {
            sequence.len()
        } * repeats;

        length += sequence;
        index += marker.end() + sequence_length;
    }
    length
}

fn part1(data: Input) -> Output {
    solve(data, false)
}

fn part2(data: Input) -> Output {
    solve(data, true)
}

#[test]
fn default() {
    let input = get_input(16, 9).unwrap();
    assert_eq!(110346, part1(&input));
    assert_eq!(10774309173, part2(&input));
}

// Input parsed (28μs)
// 1. 110346 (293μs)
// 2. 10774309173 (138μs)
// Total: 462μs